Allow interface materials to retrieve AD neighbor material properties

lindsayad opened the pull request:
Greater support for AD for interface objects which @humrickhouse needs

Commit:
Make InterfaceMaterial work with AD variables.

Closes #22296

lindsayad committed:
Create ADPenaltyInterfaceDiffusion

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1192082/) on 1236f65 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/22295/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1192095/) on 1236f65 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>549eee</th>
      <th colspan=4>#22295 1236f6</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.29%</td>
      <td>84.28%</td>
      <td>-0.01%</td>
      <td>25.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>80089</td>
      <td>80100</td>
      <td>+11</td>
      <td>6</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14928</td>
      <td>14941</td>
      <td>+13</td>
      <td>18</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22295/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22295/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/22295/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/22295/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/22295/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/22295/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/22295/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/22295/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/22295/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/22295/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/22295/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/22295/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/22295/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/22295/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/22295/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/22295/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/22295/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/22295/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/22295/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/22295/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/22295/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/22295/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/22295/coverage/richards)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/22295/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/22295/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/22295/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/22295/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/22295/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 25.00% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad committed:
Create GenericInterfaceKernel

- Leverage it in a generic PenaltyInterfaceDiffusion
- Create a test for the AD version of PenaltyInterfaceDiffusion

This commit also creates an instantiations file to make sure I can
be confident we are covering templated methods

lindsayad committed:
Add missing coverage of getNeighborADMaterialProperty APIs

moosebuild commented:
Job [Nonsparse AD](https://civet.inl.gov/job/1209975/) on 5d21b68 : invalidated by @lindsayad

lindsayad commented:
@cticenhour would you be willing to review this when you have a chance? No rush

cticenhour commented:
No problem - I will add it to my list!

cticenhour commented on test/tests/interfacekernels/1d_interface/tests:
```diff
@@ -1,45 +1,55 @@
+      input = 'ad-coupled-value-coupled-flux.i'
+      exodiff = 'ad-coupled-value-coupled-flux_out.e'
+      detail = 'using a Jacobian computed via automatic differentiation'
+      ad_indexing_type = 'global'
```
Is it important to do the indexing globally here due to the interface itself?

lindsayad replied:
No this is because I use a variable functor in the test and the variable functor is only supported by global AD indexing. AD interface kernels alone work with both AD indexing types

cticenhour reviewed (changes requested):
Good overall - see comments below.

lindsayad committed:
Apply Casey review

- Capitalization
- Remove PHW comment lines

Co-authored-by: Casey Icenhour <cticenho@ncsu.edu>

lindsayad commented:
thanks @cticenhour for the review!
